source/geod24/LocalRest.d

```diff
@@ -419,27 +430,59 @@ public final class RemoteAPI (API) : API
             }
         }
 
+        // if the node is "sleeping", but doesn't want to drop messages,
+        // it should store them here. we need to track the queue of messages
+        // on our own, to avoid accidentally dropping priority messages in
+        // receiveTimeout() as in the previous implementation.
+        Response[] response_queue;
+        Command[] command_queue;
+
         try scheduler.start(() {
                 bool terminated = false;
                 while (!terminated)
                 {
                     C.receiveTimeout(10.msecs,
                         (C.OwnerTerminated e) { terminated = true; },
                         (TimeCommand s)      {
-                            Thread.sleep(s.dur);
-                            if (s.drop)
-                                removeMessages(size_t.max);
+                            drop_messages = s.drop;
+                            sleep_until = Clock.currTime + s.dur;
                         },
                         (FilterAPI filter_api) {
                             filter = filter_api;
                         },
                         (Response res) {
-                            handle(res);
+                            if (!isSleeping())
+                            {
+                                handle(res);
+                            }
+                            else
+                            {
+                                if (!drop_messages)
+                                    response_queue ~= res;
+                            }
                         },
                         (Command cmd)
                         {
-                            handle(cmd);
+                            if (!isSleeping())
+                            {
+                                handle(cmd);
+                            }
+                            else
+                            {
+                                if (!drop_messages)
+                                    command_queue ~= cmd;
+                            }
                         });
+
+                    // now handle any leftover messages after any sleep() call
+                    if (!isSleeping())
+                    {
+                        response_queue.each!(res => handle(res));
+                        response_queue.length = 0;
+
+                        command_queue.each!(res => handle(res));
+                        command_queue.length = 0;
```

Comment:
There's a problem with this: Messages need to be processed in the order they are received

Reply:
That's a good point.